File: services/notifications-service/src/infrastructure/messaging/event_consumer.py
"""
=============================================================================
CAPA DE INFRAESTRUCTURA - Event Consumer (Notifications)
=============================================================================
Consume eventos de tickets desde RabbitMQ y genera notificaciones.

PATRÓN: Event-Driven Architecture
  Este es el corazón del notifications-service.
  No recibe requests HTTP para generar notificaciones.
  Todo se dispara por EVENTOS del Event Bus.

FLUJO:
  1. tickets-service publica ticket.created en RabbitMQ
  2. RabbitMQ enruta a la queue notifications.ticket_events
  3. Este consumer recibe el mensaje
  4. Crea una notificación en la BD
  5. "Envía" la notificación (log en el lab, email en producción)

IDEMPOTENCIA:
  Cada notificación registra el source_event_id (ID del evento origen).
  Si el mismo evento llega dos veces (ej: reintento), se ignora el duplicado.
  Esto es crítico en sistemas distribuidos donde los mensajes pueden duplicarse.

MÚLTIPLES INSTANCIAS:
  Si hay N instancias de notifications-service, RabbitMQ distribuye
  los mensajes entre ellas (round-robin). Cada evento lo procesa UNA instancia.
=============================================================================
"""

import json
import logging
import os
from uuid import UUID

import aio_pika

from src.infrastructure.database.connection import AsyncSessionLocal
from src.infrastructure.database.models import NotificationModel

logger = logging.getLogger(__name__)

RABBITMQ_URL = os.getenv("RABBITMQ_URL", "amqp://guest:guest@localhost:5672/")
SERVICE_NAME = os.getenv("SERVICE_NAME", "notifications-service")

# Templates de notificación por tipo de evento (Ubiquitous Language)
NOTIFICATION_TEMPLATES = {
    "ticket.created": {
        "subject": "Tu ticket #{id} ha sido recibido",
        "body": "Hola {customer_name}, tu ticket '{subject}' ha sido recibido. "
                "Te notificaremos cuando un agente sea asignado.",
        "recipient_type": "CUSTOMER",
    },
    "ticket.assigned": {
        "subject": "Tu ticket #{id} ha sido asignado",
        "body": "Hola {customer_name}, el agente {assigned_agent_name} ha sido asignado "
                "a tu ticket '{subject}'. Pronto recibirás una respuesta.",
        "recipient_type": "CUSTOMER",
    },
    "ticket.resolved": {
        "subject": "Tu ticket #{id} ha sido resuelto",
        "body": "Hola {customer_name}, tu ticket '{subject}' ha sido resuelto. "
                "Si el problema persiste, puedes reabrir el ticket.",
        "recipient_type": "CUSTOMER",
    },
    "ticket.closed": {
        "subject": "Tu ticket #{id} ha sido cerrado",
        "body": "Hola, tu ticket #{id} ha sido cerrado. Gracias por contactarnos.",
        "recipient_type": "CUSTOMER",
    },
}


async def start_consuming():
    """Conecta a RabbitMQ y comienza a consumir eventos de tickets."""
    try:
        connection = await aio_pika.connect_robust(RABBITMQ_URL)
        channel = await connection.channel()

        # Prefetch=1: procesar un mensaje a la vez (distribución equitativa entre instancias)
        await channel.set_qos(prefetch_count=1)

        # Consumir de la queue pre-configurada
        queue = await channel.get_queue("notifications.ticket_events")
        await queue.consume(process_ticket_event)

        logger.info("Started consuming from notifications.ticket_events")

    except Exception as e:
        logger.error(f"Failed to start consuming: {e}")


async def process_ticket_event(message: aio_pika.IncomingMessage):
    """
    Procesa un evento de ticket y genera la notificación correspondiente.

    Implementa IDEMPOTENCIA: si el evento ya fue procesado (source_event_id
    duplicado), se ignora silenciosamente.
    """
    async with message.process():
        try:
            event = json.loads(message.body)
            event_type = event.get("event_type", "")
            payload = event.get("payload", {})
            event_id = event.get("event_id")
            correlation_id = event.get("correlation_id", "")

            logger.info(
                f"Processing event: {event_type}",
                extra={"event_id": event_id, "correlation_id": correlation_id},
            )

            # Verificar si tenemos template para este tipo de evento
            template = NOTIFICATION_TEMPLATES.get(event_type)
            if not template:
                logger.warning(f"No template for event type: {event_type}")
                return

            # Generar contenido de la notificación usando el template
            short_id = payload.get("id", "")[:8]
            subject = template["subject"].format(id=short_id, **payload)
            body = template["body"].format(id=short_id, **{
                "customer_name": payload.get("customer_name", "Cliente"),
                "subject": payload.get("subject", ""),
                "assigned_agent_name": payload.get("assigned_agent_name", ""),
            })

            # Persistir la notificación
            async with AsyncSessionLocal() as session:
                notification = NotificationModel(
                    event_type=event_type,
                    recipient_id=UUID(payload.get("customer_id")) if payload.get("customer_id") else None,
                    recipient_type=template["recipient_type"],
                    recipient_email=payload.get("customer_email"),
                    channel="INTERNAL",
                    subject=subject,
                    body=body,
                    status="SENT",  # En el lab, siempre "enviado" (simulado)
                    source_event_id=UUID(event_id) if event_id else None,
                )
                session.add(notification)
                await session.commit()

            # En el lab, logueamos la "notificación enviada".
            # En producción, aquí se llamaría al servicio de email/SMS/push.
            logger.info(
                f"Notification sent: {subject}",
                extra={
                    "recipient_email": payload.get("customer_email"),
                    "event_type": event_type,
                    "correlation_id": correlation_id,
                },
            )

        except Exception as e:
            logger.error(f"Error processing event: {e}")

File: services/agents-service/src/infrastructure/messaging/event_consumer.js
/**
 * =============================================================================
 * Event Consumer para agents-service
 * =============================================================================
 * Consume eventos de tickets para mantener sincronizado:
 *   - active_tickets_count: se incrementa con ticket.assigned,
 *     se decrementa con ticket.resolved / ticket.closed
 *
 * CONSISTENCIA EVENTUAL:
 *   El contador puede estar brevemente desincronizado con la realidad
 *   en tickets-service. Esto es aceptable y esperado en microservicios.
 *
 * MÚLTIPLES INSTANCIAS:
 *   Todas las instancias comparten la misma queue (agents.ticket_events).
 *   RabbitMQ distribuye mensajes en round-robin entre instancias.
 *   Como las operaciones son sobre la BD compartida, no hay conflicto.
 * =============================================================================
 */

const amqplib = require('amqplib');
const { pool } = require('../database/connection');

const RABBITMQ_URL = process.env.RABBITMQ_URL || 'amqp://guest:guest@localhost:5672/';
const SERVICE_NAME = process.env.SERVICE_NAME || 'agents-service';

async function startConsuming() {
  try {
    const connection = await amqplib.connect(RABBITMQ_URL);
    const channel = await connection.createChannel();
    await channel.prefetch(1);

    await channel.consume('agents.ticket_events', async (msg) => {
      try {
        const event = JSON.parse(msg.content.toString());
        const { event_type, payload, correlation_id } = event;

        console.log(JSON.stringify({
          timestamp: new Date().toISOString(), service: SERVICE_NAME, level: 'INFO',
          message: `Received event: ${event_type}`, correlation_id,
        }));

        if (event_type === 'ticket.assigned' && payload.assigned_agent_id) {
          // Incrementar contador del agente asignado
          await pool.query(
            `UPDATE agents SET active_tickets_count = active_tickets_count + 1,
             status = CASE WHEN active_tickets_count + 1 >= max_tickets THEN 'BUSY' ELSE status END,
             updated_at = NOW()
             WHERE id = $1`,
            [payload.assigned_agent_id]
          );
          console.log(JSON.stringify({
            timestamp: new Date().toISOString(), service: SERVICE_NAME, level: 'INFO',
            message: `Agent ${payload.assigned_agent_id}: incremented active tickets`, correlation_id,
          }));
        }

        if ((event_type === 'ticket.resolved' || event_type === 'ticket.closed') && payload.assigned_agent_id) {
          // Decrementar contador del agente
          await pool.query(
            `UPDATE agents SET active_tickets_count = GREATEST(active_tickets_count - 1, 0),
             status = CASE WHEN status = 'BUSY' AND active_tickets_count - 1 < max_tickets THEN 'ONLINE' ELSE status END,
             updated_at = NOW()
             WHERE id = $1`,
            [payload.assigned_agent_id]
          );
          console.log(JSON.stringify({
            timestamp: new Date().toISOString(), service: SERVICE_NAME, level: 'INFO',
            message: `Agent ${payload.assigned_agent_id}: decremented active tickets`, correlation_id,
          }));
        }

        channel.ack(msg);
      } catch (err) {
        console.error(JSON.stringify({
          timestamp: new Date().toISOString(), service: SERVICE_NAME, level: 'ERROR',
          message: `Error processing ticket event: ${err.message}`,
        }));
        channel.nack(msg, false, false);
      }
    });

    console.log(JSON.stringify({
      timestamp: new Date().toISOString(), service: SERVICE_NAME, level: 'INFO',
      message: 'Started consuming from agents.ticket_events',
    }));

    connection.on('close', () => setTimeout(startConsuming, 5000));
  } catch (err) {
    console.error(JSON.stringify({
      timestamp: new Date().toISOString(), service: SERVICE_NAME, level: 'ERROR',
      message: `Consumer connect failed: ${err.message}`,
    }));
    setTimeout(startConsuming, 5000);
  }
}

module.exports = { startConsuming };
